### ISSUE TITLE
Low GPU utilization

### ISSUE THREAD

**Issue by Rizhiy** [roytseng-tw/Detectron.pytorch#8]
I'm training on 4 GPUs with 8 workers but getting only about 50% GPU utilization.

What can be the problem?

**Comment by roytseng-tw** [comment]
What's your CPU model ?

**Comment by roytseng-tw** [comment]
Your training time for 100 iterations seems normal. BTW, which architecture did you train ?
I know that gpu utilization is a little bumpy during the training time. However, the overall training speed is similar to Detectron's in Caffe2.

**Comment by roytseng-tw** [comment]
Possible way to further improve the gpu utilization:
1. Use coalesced broadcast for dataloader
2. Move entries in rpn returned blobs to the target gpu at once.

**Comment by roytseng-tw** [comment]
Official DataParallel will broadcast all the input Variables to GPUs. However, many rpn related computations are done in CPU, and it's unnecessary to put those related inputs on GPUs. So I make DataParallel be capable of keeping certain keyword inputs on CPU.

Besides, I also make returned value of type dictionary work with DataParallel.

**Comment by Rizhiy** [comment]
Didn't @jwyang change all RPN code to use torch operations, so that it can be done completely on GPU?

I am just wondering whether the transfer between CPU and GPU can be the bottleneck.

**Comment by roytseng-tw** [comment]
I use the RPN implementation from Detectron so as to 
1) easily migrate different network arichectures from Detectron
2) reproduce the same results directly from Detectron's pretrained weights by ensuring that all computational logic are the same

Detectron implements those RPN computations in caffe2 python layer, and hence the computation is also on CPU.
Actually in this pytorch implementation, comparing to Detectron, I save the time to move some of the input variables from GPU back to CPU by the customized DataParalle module. Besides, I only move necessary outputs from RPN to GPU.

I think, in general, I transfer less amount of data between GPU and CPU.
Moreover, when measuring the actual training time, it's almost comparable to Detectron's. So, I infer that my implementation should have similar gpu utilization.
In fact, I wonder that if nvidia-smi's gpu utilization stats is a fair/effective measurement across different deep learning frameworks.

**Comment by JackHenry1992** [comment]
Did you get the right loss when use 4 GPUs? As #19 , I get the 'loss=NAN' when training with 4 GPUs.

**Comment by Rizhiy** [comment]
@JackHenry1992 Yes, I have training working with 4 GPUs.

**Comment by roytseng-tw** [comment]
closing for inactivity
